### PR TITLE
INT-4101 MessagingException to include failedMessage

### PR DESCRIPTION
**Pull request by m-terra**
JIRA: https://jira.spring.io/browse/INT-4101

### CODE REVIEW

File: spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java]: I think, I'm fine with the solution.
I thought that we don't that `ThreadLocal` when we have a TX, but indeed it is already finished and cleaned on the `Poller.this.pollingTask.call()`. And if TX throws exception during commit, we really don't have a TX-synched resource to consult for the `Message`.

So, LGTM here, although I still think that it should be `MessageHandlingException`. Just because it isn't so generic case, but really the `handleMessage(Message)` problem.

File: spring-integration-core/src/test/java/org/springframework/integration/endpoint/PollingEndpointErrorHandlingTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/endpoint/PollingEndpointErrorHandlingTests.java]: Would be better do not exceed `120` line length.

**Review by artembilan — CHANGES_REQUESTED** [review]
A couple simple concerns.

@garyrussell , the solution looks so simple, that I worry I'm missing something.

LGTM, but would like to see your opinion.
Thanks

File: spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java]: Actually, I don't think `MessageHandlingException` is correct here; if the exception occurred on a downstream handler, it will already be an `MHE`; it not, then it's not really an MHE it's an "unknown" type of ME.

I wonder if we should take the opportunity we discussed last week - i.e. an enhanced messaging exception where the payload contains the failedMessage as well as this (`inputMessage`).

**Comment by garyrussell** [comment]
Overall, it looks ok, but I don't like the overhead of maintaining a `ThreadLocal` on every poll, for every poller since it's only when a non-messaging-aware advice (such as TxAdvice) throws an exception that we need to create the enhanced exception. That's why I originally suggested adding an advice to the advice chain in the JIRA.

This branch has a non-linear commit log; it needs to be rebased to master (and conflicts resolved).

**Comment by andybaer** [comment]
I don't like to use ThreadLocal's as well but we needed a fix and this one did it. The advice approach I out of my league ;)
I did not notice the conflicts and I can change the formatting of course. Please let me know if I should proceed.

File: spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java]: Well, that is fully different story (enhanced exception) and let's consider it in the separate task!

And well, let it be just general `ME`!

**Comment by artembilan** [comment]
Well another suggestion is something like end-user choice - a `MethodInterceptor` implementation which will be applied around `TxAdvice` in the `AbstractPollingEndpoint` and wrap an exception from there to the `MessagingException`. But still... how to obtain a `Message` there?

If we apply an advice just to the `handleMessage()`, we will be already within `TxAdvice`, therefore won't be able to interfere to the TX exception around.

I'm confused how you are going to proceed without `ThreadLocal`...

**Comment by garyrussell** [comment]
>Well, that is fully different story (enhanced exception) and let's consider it in the separate task!

Well, if we agree that will be useful (I agree it's a separate issue), then we will need the `ThreadLocal` anyway so let's go ahead and polish and merge this - @artembilan you or me?

Thanks @andybaer !!

**Comment by andybaer** [comment]
I rebased and reformatted. Thanks to both of you for considering this fix!

**Comment by garyrussell** [comment]
>I'm confused how you are going to proceed without ThreadLocal...

The TL would be in the advice instead, and therefore only used when the advice is configured - we can still consider some optimization later, perhaps before GA; let's go with it as-is for now and revisit if/when we consider the enhanced ME.

**Comment by garyrussell** [comment]
>I'm confused how you are going to proceed without ThreadLocal...

The message would be a local variable in the advice instead (captured before the `proceed()`) - we can still consider some optimization later, perhaps before GA; let's go with it as-is for now and revisit if/when we consider the enhanced ME.

**Comment by andybaer** [comment]
Had some trouble rebasing but now the commit history is linear

**Comment by artembilan** [comment]
Guys, looking to the problem one more time, I don't see (yet) some generic solution.
Let's fix the current problem and come back if we will figure out what is needed.

So, as we know the `handleMessage(Message)` throws `MessagingException` with the "guilty" message.
Therefore we're good in the `errorChannel` flow and have some message context.

When we come to the TX and error during commit, we are already far away from the `handleMessage(Message)` and just lose the "guilty" message context.
Fortunately we bind to the TX there `IntegrationResourceHolder` via `IntegrationResourceHolderSynchronization`, which has an option like `shouldUnbindAtCompletion()` with `true` by default.
We can switch it into `false` and unbind manually in the same place like @andybaer does it now.
WDYT?

**Comment by andybaer** [comment]
Sound like a nice solution. This implies a public setter for `shouldUnbindAtCompletion()`. I can change it to this implementation if you agree.

**Comment by garyrussell** [comment]
Sounds good to me, but it will require a `TransactionSynchronizationFactory` configured into the poller (that's when we save off the message).

Perhaps we could always configure a pass-through one on all transactional pollers?

**Comment by artembilan** [comment]
Well, that's true.

So, the way to go:
1. Add setter to the `IntegrationResourceHolderSynchronization`  to let to configure `shouldUnbindAtCompletion` and set it in the `AbstractPollingEndpoint` to `false`.
2. Use a new `catch` logic in the `Poller` class to extract `IntegrationResourceHolder` from the `TransactionSynchronizationManager` and `unbind` it in the `finally`
3. Introduce new `PassThroughTransactionSynchronizationFactory` to use when there is no provided `transactionSynchronizationFactory` and `TransactionSynchronizationManager.isActualTransactionActive()`.

This way we nail the premise of the problem after TX commit exception and we won't have an extra `ThreadLocal` variable.

The long way solution might be like `exposeInputMessage = true/false` on the `AbstractPollingEndpoint` and introduce something like `InputMessageContext` with the `ThreadLocal` already.
That `InputMessageContext` with `public static` methods can be used in any downstream component. 
Exactly similar way how we use these days `SecurityContextHolder`.

Any other thoughts?
Thanks

**Comment by andybaer** [comment]
This sounds very interesting. Let me give it a shot and I will come back on the same pull request.

Thanks a lot @artembilan & @garyrussell !

**Comment by andybaer** [comment]
Hi @artembilan 

I gave it a try but `TransactionSynchronizationManager.clearSynchronization()` and `TransactionSynchronizationManager.clear()` both are invoked before the exception arrives in the poller's catch block. So the `shouldUnbindAtCompletion` setting is not preventing the advice from clearing the tx state. Code is on this branch: https://github.com/andybaer/spring-integration/tree/INT-4101-NEW

Any advice how to proceed?

Thanks!

**Comment by artembilan** [comment]
@andybaer ,

thank you for the effort!

Well, even if `TransactionSynchronizationManager` is cleared that way in the end of transaction, pay attention that `resources` isn't cleared. So, we still have there `IntegrationResourceHolder` with a desired message. We should do `TransactionSynchronizationManager.unbindResource(this.resourceKey);` manually any way. Just because it is correct - we don't have TX any more, but we still pollute `ThreadLocal` .

**Comment by artembilan** [comment]
OK. I see your problem.
You should use `TransactionSynchronizationManager.getResource()` instead of iteration over `TransactionSynchronizationManager.getSynchronizations()`

**Comment by andybaer** [comment]
Yes thanks your comment was clear. Just running all the tests and rebase and then I will push to this pull request

**Comment by andybaer** [comment]
@artembilan 

The new implementation using `TransactionSynchronizationManager` is finished. Please have a look.

Thanks!

File: spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java]: We only want to take the hit of saving the message in a TL __if the poller is transactional__ - we can easily tell that in the XML parser and via the DSL spec, but not so easily via the `@Poller` annotation; we may need to add a `transactional` property to `@Poller` for that.

On the other hand, any "around" advice that can throw an exception may need access to the message if it throws the exception after the proceed.

So, I don't mind if we always apply it if the advice chain is > 0 length - that's probably the simplest solution that would work in all cases while only taking the extra unneeded hit if someone configures an advice chain.

Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java]: Well, my idea was to fix exactly the TX case. Therefore this `PassthroughTransactionSynchronizationFactory`.
We don't care about transactional poller or not but just use `TransactionSynchronizationManager.isActualTransactionActive()`.
So, that still should be enough for us here as well.
Any other "exotic" advices can access to the message in their after `proceed()` block via the same `ThreadLocal` solution but already as a population in the downstream flow.

Let me know if I have missed something or you need some other explanation.

File: spring-integration-core/src/main/java/org/springframework/integration/transaction/IntegrationResourceHolderSynchronization.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/transaction/IntegrationResourceHolderSynchronization.java]: `true` by default.
That is only `AbstractPollingEndpoint` to take care about the `TL` by its own.

File: spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java]: Minor, but we can extract local variable for the `getResourceToBind()`.
Not sure if compiler does an optimization.
Although it can't because of possibility to have a different result from each method invocation.

File: spring-integration-core/src/main/java/org/springframework/integration/transaction/PassthroughTransactionSynchronizationFactory.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/transaction/PassthroughTransactionSynchronizationFactory.java]: New classes only with the current year.

File: spring-integration-core/src/test/java/org/springframework/integration/dispatcher/FailingCommitTransactionManager.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/dispatcher/FailingCommitTransactionManager.java]: M-m-m. This looks like a test-case scope.
Let's move it into the test class - `PollingTransactionTests`!

File: spring-integration-core/src/test/java/org/springframework/integration/dispatcher/PollingTransactionTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/dispatcher/PollingTransactionTests.java]: Can't we combine two these new tests?
I'm very sensible to any build time increases, like the same app ctx start and close. 😄 

Well, I see that is a style of this test class at all.
Would be better to revise it to a managed test with a single ctx for better performance.
But different story...
OK, let's leave your code as is for that future refactoring.

File: spring-integration-core/src/test/java/org/springframework/integration/endpoint/PseudoTransactionalMessageSourceTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/endpoint/PseudoTransactionalMessageSourceTests.java]: M-m-m. What made you to do these changes?
Thanks

File: spring-integration-core/src/test/java/org/springframework/integration/endpoint/PseudoTransactionalMessageSourceTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/endpoint/PseudoTransactionalMessageSourceTests.java]: `TestUtils.createTestApplicationContext()` is there for you.
OTOH you always can do `mock(BeanFactory.class)` if that satisfies a requirement.

File: spring-integration-test/src/main/java/org/springframework/integration/test/util/TestUtils.java
Comment (artembilan) [inline spring-integration-test/src/main/java/org/springframework/integration/test/util/TestUtils.java]: As I said: `TestUtils.createTestApplicationContext()` so, no changes here, please.

File: spring-integration-core/src/test/java/org/springframework/integration/dispatcher/PollingTransactionTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/dispatcher/PollingTransactionTests.java]: Usually our CI machine is pretty slow, so 3 secs might not be enough to wait for the answer from another thread.
Right, I see that it is a practice of this test class, but let's move on to the more robust solution.
So, 10 sec is our typical time.

File: spring-integration-core/src/main/java/org/springframework/integration/transaction/PassthroughTransactionSynchronizationFactory.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/transaction/PassthroughTransactionSynchronizationFactory.java]: M-m-m. I don't like "stub" word. Let's say "A simple TransactionSynchronizationFactory implementation which creates blah-blah and register blah-blah"

"Register" is very important word here because we do `TransactionSynchronizationManager.bindResource()` unconditionally.

**Review by artembilan — CHANGES_REQUESTED** [review]
That's all for this round.
And, to be honest, it is so organic solution that I don't see reason to document it somehow...

File: spring-integration-core/src/main/java/org/springframework/integration/transaction/IntegrationResourceHolderSynchronization.java
Comment (andybaer) [inline spring-integration-core/src/main/java/org/springframework/integration/transaction/IntegrationResourceHolderSynchronization.java]: oops

File: spring-integration-core/src/test/java/org/springframework/integration/endpoint/PseudoTransactionalMessageSourceTests.java
Comment (andybaer) [inline spring-integration-core/src/test/java/org/springframework/integration/endpoint/PseudoTransactionalMessageSourceTests.java]: This test did not bind resources to the `TransactionSynchronizationManager` before my changes because the `TransactionSynchronizationFactory` was null. Now that it does bind resources they must be cleared otherwise subsequent tests fail.

File: spring-integration-core/src/test/java/org/springframework/integration/endpoint/PseudoTransactionalMessageSourceTests.java
Comment (andybaer) [inline spring-integration-core/src/test/java/org/springframework/integration/endpoint/PseudoTransactionalMessageSourceTests.java]: Personal preference, I try to avoid mocking ;)

File: spring-integration-core/src/test/java/org/springframework/integration/dispatcher/PollingTransactionTests.java
Comment (andybaer) [inline spring-integration-core/src/test/java/org/springframework/integration/dispatcher/PollingTransactionTests.java]: I agree, no big change

**Comment by andybaer** [comment]
All checked in, thanks for the review!

Not sure about the consent on when the presence of `PassthroughTransactionSynchronizationFactory`

File: spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java]: `TSM.iATA()` accesses a thread local - since 99.99% of pollers are __not transactional__ I'd prefer to not take two TL hits (here and in the poller's finally block) on every poll - a simple `null` check on the TSF field will be much more efficient.

Something less than 99.99%, but probably > 98% of pollers don't have an advice chain at all, so I can live with that (presence of an advice chain) being the condition we use to decide whether or not to store the message in a TL.

@andybaer You don't need to take any action until @artembilan and I have resolved our differences 😄

Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java]: OK, Gary! 

That is fully different story. Sorry, that wasn't obvious by your comment that you wouldn't like to hit `ThreadLocal` 😄 

Does that mean that we have to revise all our `TSM.iATA()` usage
For example `ConnectionFactoryUtils`:
```
public static RabbitResourceHolder bindResourceToTransaction(RabbitResourceHolder resourceHolder,
		ConnectionFactory connectionFactory, boolean synched) {
	if (TransactionSynchronizationManager.hasResource(connectionFactory)
			|| !TransactionSynchronizationManager.isActualTransactionActive() || !synched) {
		return (RabbitResourceHolder) TransactionSynchronizationManager.getResource(connectionFactory);
	}
```
And `ImapIdleChannelAdapter`:
```
if (TransactionSynchronizationManager.isActualTransactionActive()) {
		if (ImapIdleChannelAdapter.this.transactionSynchronizationFactory != null) {
```
?

I don't mind about optimization. 
So, another trick is check if our `adviceChain` contains `TransactionInterceptor`.
If it is there and `TSF == null` we populate this our new `PassthroughTransactionSynchronizationFactory`

Does that work for you?

File: spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java]: There is one more in this method - `TransactionSynchronizationManager.getResource(getResourceToBind())`

File: spring-integration-core/src/main/java/org/springframework/integration/transaction/PassthroughTransactionSynchronizationFactory.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/transaction/PassthroughTransactionSynchronizationFactory.java]: Would be better to say `{@link IntegrationResourceHolder} under provided {@code key}` instead.
Plus the general class JavaDoc must end with a period.

File: spring-integration-core/src/main/java/org/springframework/integration/transaction/IntegrationResourceHolderSynchronization.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/transaction/IntegrationResourceHolderSynchronization.java]: Since you did some changes in this class, add your name to the `@author` and update Copyright to the actual year. In all affected classes

File: spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java]: We should probably take a look at other uses but I am not too concerned about, say, ImapIdle but, in this case, we would have added unnecessary overhead to a zillion pollers worldwide.

As I said, I don't mind if the test is for a non-empty advice chain, but if you want to further optimize the test to look for a tx interceptor, I am ok with that too. The only problem with that would be if someone wraps the tx interceptor for some reason.

File: spring-integration-core/src/test/java/org/springframework/integration/endpoint/PseudoTransactionalMessageSourceTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/endpoint/PseudoTransactionalMessageSourceTests.java]: No `e.printStackTrace()`, please, at all.
It doesn't provide much info and just pollutes logs on CI builds

File: spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java]: Well, for that purpose we can document:
"We have `PassthroughTransactionSynchronizationFactory` if you would like to have `TL` value for message in case of custom TX Advice. It is applied by default if TX Advice is exactly `TransactionInterceptor`."

**Comment by artembilan** [comment]
@andybaer ,

Please, stop to `squash` commits. It is much easier for us to review just small peaces of the changes.
Plus now we have lost all the history of the changes.

**Comment by andybaer** [comment]
Sorry @artembilan

I thought that was prefered since there were many usless commits. But I see your point. No more squashing! :)

**Comment by artembilan** [comment]
OK, @andybaer , now we are very close to the merge.
Please, try to address the rest couple my concerns and add a not into `transactions.adoc` about this our new `PassthroughTransactionSynchronizationFactory`.

@garyrussell , any other concerns from your side?

Thanks both for so comprehensive team work! 😄

**Comment by andybaer** [comment]
@artembilan 

I addressed everything else so I will put my hands on the adoc

**Comment by garyrussell** [comment]
LGTM; thanks both!

File: spring-integration-core/src/main/java/org/springframework/integration/transaction/IntegrationResourceHolderSynchronization.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/transaction/IntegrationResourceHolderSynchronization.java]: Isn't addressed.

File: spring-integration-core/src/main/java/org/springframework/integration/transaction/PassthroughTransactionSynchronizationFactory.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/transaction/PassthroughTransactionSynchronizationFactory.java]: DITTO.

I can do that on merge any way, but since we are expecting one more commit from you about docs, it would be better to have this there as well.
Thanks

File: spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java]: No, you have to leave the first year and fix the last to the current.
The first year is about the birth date, the last about maturity.

File: src/reference/asciidoc/transactions.adoc
Comment (artembilan) [inline src/reference/asciidoc/transactions.adoc]: Double `a` and it really must be a separate paragraph.

Plus it would be helpful to say that now `AbstractPollingEndpoint` correctly throws `MessagingException` with the polled message as `failedMessage` property. And say that it is picked up from the `TransactionSynchronizationManager`.
Having a note about `PassthroughTransactionSynchronizationFactory` confuses me a bit now. 
The question is: what is the purpose of this factory used by default?

Let me know if I'm not clear yet and I'll come back to you with the polishing as I see for review and other comments.

Comment (andybaer) [inline src/reference/asciidoc/transactions.adoc]: I'm pretty sure I understand but let's see...

**Review by artembilan — APPROVED** [review]

**Comment by artembilan** [comment]
Awesome, @andybaer !

Please, review the polishing before I proceed with the merge: https://github.com/artembilan/spring-integration/commit/20b0888d2226059733609258113d4f8d4cae405b

**Comment by garyrussell** [comment]
I made some comments on the polish.

Many thanks, once again @andybaer and @artembilan

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/3aa830b7c0c8cde3eba735cbd33563e180bf334d

@andybaer ,

thank you for your patience for so complex problem and contribution, of course.
Looking forward for more!

**Comment by andybaer** [comment]
Thanks @artembilan and @garyrussell !!!

It's a great honour as well as free lessons from real professionals ;)

I'd be up for more, guess I will browse the Jira then?!

**Comment by artembilan** [comment]
Yep! Sure thing! There is always something open to do 😄 

Plus feel free to come to us with any other ideas for improvements or just bugs.
Through the same JIRA, of course! 😉

**Pull request closed without merging**
